org.eclipse.xtext.builder.standalone/src/org/eclipse/xtext/builder/standalone/StandaloneBuilderState.java

```diff
@@ -0,0 +1,563 @@
+		syncOutputFiles(this.stubFiles, currentStubFiles);
+	}
+
+	private void syncOutputFiles(Map<IPath, HashCode> originalFiles, Map<IPath, HashCode> newFiles) {
+		MapDifference<IPath, HashCode> difference = Maps.difference(originalFiles, newFiles);
+
+		// Handle all files that are expected to be there but missing
+		for (IPath missing : difference.entriesOnlyOnLeft().keySet()) {
+			discardOutputAndStubData(outputToInputFile.remove(missing));
+			discardOutputAndStubData(stubToInputFile.remove(missing));
+		}
+		// Handle all files that are present but different
+		for (IPath different : difference.entriesDiffering().keySet()) {
+			discardOutputAndStubData(outputToInputFile.remove(different));
+			discardOutputAndStubData(stubToInputFile.remove(different));
+		}
+	}
+
+	private void discardOutputAndStubData(URI inputFile) {
+		if (inputFile != null && sourceFiles.put(inputFile, null) != null) {
+			IPath[] outputFiles = inputToOutputFiles.remove(inputFile);
+			if (outputFiles != null) {
+				for (IPath outputFile : outputFiles) {
+					outputFile.toFile().delete();
+					outputToInputFile.remove(outputFile);
+					this.outputFiles.remove(outputFile);
+				}
+			}
+			IPath[] stubFiles = inputToStubFiles.remove(inputFile);
+			if (stubFiles != null) {
+				for (IPath stubFile : stubFiles) {
+					stubFile.toFile().delete();
+					stubToInputFile.remove(stubFile);
+					this.stubFiles.remove(stubFile);
+				}
+			}
+		}
+	}
+
+	IResourceDescription.Event sourceChanges(List<URI> resourceURIs, Collection<URI> newResources) {
+		Map<URI, HashCode> newSourceFiles = new HashMap<>();
+		for (URI sourceResource : resourceURIs) {
+			newSourceFiles.put(sourceResource, hash(sourceResource));
+		}
+		return diffDslResources(sourceFiles, newSourceFiles, newResources);
+	}
+
+	/**
+	 * Returns the delta of deleted resources and an accumulation of newly
+	 * available resources.
+	 */
+	IResourceDescription.Event libraryChanges(List<URI> resourceURIs, Collection<URI> newResources) {
+		Map<URI, HashCode> newBinaryFiles = new HashMap<>();
+		for (URI binaryResource : resourceURIs) {
+			newBinaryFiles.put(binaryResource, hash(binaryResource));
+		}
+		return diffDslResources(libraryFiles, newBinaryFiles, newResources);
+	}
+
+	private IResourceDescription.Event diffDslResources(Map<URI, HashCode> originalFiles, Map<URI, HashCode> newFiles,
+			Collection<URI> accumulator) {
+		MapDifference<URI, HashCode> difference = Maps.difference(originalFiles, newFiles);
+		List<IResourceDescription.Delta> deltas = new ArrayList<>();
+		for (URI removed : difference.entriesOnlyOnLeft().keySet()) {
+			IResourceDescription oldDescription = index.getResourceDescription(removed);
+			if (oldDescription != null) {
+				deltas.add(new DefaultResourceDescriptionDelta(oldDescription, null));
+				index.removeDescription(removed);
+			}
+			issues.removeAll(removed);
+			IPath[] outputFiles = inputToOutputFiles.remove(removed);
```

Comment:
can this handle file renames with same generated files?

Reply:
Great point. I'll add a testcase and only delete if the output->input mapping still matches.

Reply:
In this particular location, I think it's fine since the renamed input file will be treated as removed and recreated.
The output file will be rewritten potentially with same content. The compile result is valid.
I'm not sure this is a common use case and worthy to be optimized further.

Reply:
Tests added, fix applied.